mgr/dashboard: Unify the look of dashboard charts

tspmelo opened the pull request:
Fixes: http://tracker.ceph.com/issues/39384

`Signed-off-by: Tiago Melo <tmelo@suse.com>`

Currently we have 3 charts in the dashboard and each one of them is different.
I would suggest to use pie chart in all and add the % to the labels.

Before:
![localhost_4200_ (1)](https://user-images.githubusercontent.com/399326/56371206-358b2b80-61ec-11e9-85c9-28fa66001649.png)

After:
![localhost_4200_](https://user-images.githubusercontent.com/399326/56371212-37ed8580-61ec-11e9-9dc9-87be720efa0c.png)


<!--
Thank you for opening a pull request!  Here are some tips on creating
a well formatted contribution.

Please give your pull request a title like "[component]: [short description]"

This is the format for commit messages:

"""
[component]: [short description]

[A longer multiline description]

Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
Signed-off-by: [Your Name] <[your email]>
"""

The Signed-off-by line is important, and it is your certification that
your contributions satisfy the Developers Certificate or Origin.  For
more detail, see SubmittingPatches.rst.

The component is the short name of a major daemon or subsystem,
something like "mon", "osd", "mds", "rbd, "rgw", etc. For ceph-mgr modules,
give the component as "mgr/<module name>" rather than a path into pybind.

For more examples, simply use "git log" and look at some historical commits.

This was just a quick overview.  More information for contributors is available here:
https://raw.githubusercontent.com/ceph/ceph/master/SubmittingPatches.rst

-->

- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

tspmelo commented:
Quest I posted in the tracker issue [1]:
> It can be hard to hover/click in a slice if that slice only has 1%, so the user would only have easy access to the value we display in the label.
Would it make more sense to display the actual value or continue to show the %??

[1] http://tracker.ceph.com/issues/39384

alfonsomthd commented:
> Question I posted in the tracker issue [1]:
> 
> > It can be hard to hover/click in a slice if that slice only has 1%, so the user would only have easy access to the value we display in the label.
> > Would it make more sense to display the actual value or continue to show the %??
> 
> [1] http://tracker.ceph.com/issues/39384

About 'Raw Capacity' chart, @pcuzner wrote (http://tracker.ceph.com/issues/38697):
"
Should the tile's title include the used+avail value (i.e. 120TB). At the moment you have to hover and do the math yourself to understand how big your cluster is.
"

tspmelo commented:
jenkins test make check arm64

julienlim commented:
+1  Using a pie chart towards consistency though there's something to consider with regards to the donut related to raw capacity.

+1 Adding % to labels, which will negate user having to click on a slice to see the %.  Furthermore, if the slice is too small, user won’t struggle to click to see the amount or %.

Regarding the Raw Capacity card, I think it’s okay to leave the label as is.  It would be good though for us to show the capacity values and not just the % as well as the total, so user would not have to perform the mental calculation to figure out the total.  E.g. 80 TB Used (80%) and 20 TB Available (20%).  Showing 100 TB Total somewhere in the white space underneath the legend would be useful.  If we kept the donut, we could show the Total in the donut.

Thoughts?

LenzGr commented:
> +1 Adding % to labels, which will negate user having to click on a slice to see the %. Furthermore, if the slice is too small, user won’t struggle to click to see the amount or %.

+1
 
> Regarding the Raw Capacity card, I think it’s okay to leave the label as is. It would be good though for us to show the capacity values and not just the % as well as the total, so user would not have to perform the mental calculation to figure out the total. E.g. 80 TB Used (80%) and 20 TB Available (20%).

+1 for showing the total - I have received requests that people would like to see the actual capacity values without having to use the mouse (e.g. when the landing page is displayed on a screen in a data center), so I'd suggest to add them to the legend as well.

> Showing 100 TB Total somewhere in the white space underneath the legend would be useful. If we kept the donut, we could show the Total in the donut.

Indeed, contrary to the other two pie charts (Client Read/Write and PG Status), the sum/aggregate value of the individual chart segments (Total Capacity) is also valuable information that should be displayed. Then again, it could also be shown in the legend next to the pie chart or in the card's title.

For reference, here's how some other Ceph UIs visualize this information:

[Calamari](https://raw.githubusercontent.com/ceph/calamari-clients/master/screenshots/Main_Dashboard.png) used a gauge that only showed the current used and total values:

![Screenshot from 2019-05-02 11-08-56](https://user-images.githubusercontent.com/263427/57065767-bf49f700-6cca-11e9-89b0-e65f79898764.png)

The [dashboard in Luminous](https://i0.wp.com/ceph.com/wp-content/uploads/2017/09/dashboard_frontpage.png) used a donut chart that displayed the percentage used in the center of the donut:

![Screenshot from 2019-05-02 11-04-56](https://user-images.githubusercontent.com/263427/57065693-86aa1d80-6cca-11e9-93d7-a11c2629f346.png)

[ceph-dash](https://raw.githubusercontent.com/crapworks/ceph-dash/master/screenshots/ceph-dash.png) also uses a gauge for percentage and shows the raw values below the widget:

![Screenshot from 2019-05-02 11-06-17](https://user-images.githubusercontent.com/263427/57065831-eef8ff00-6cca-11e9-91f8-8adb39648fbd.png)

Lastly, [Croit](https://croit.io/images/screenshots/2018-09-24_nightmode.png) uses no chart at all and just displays the percentage used and total capacity:

![Screenshot from 2019-05-02 11-13-31](https://user-images.githubusercontent.com/263427/57066024-67f85680-6ccb-11e9-89b9-1a41470b7fc7.png)

My proposal for how to modify the "Raw Capacity" card in this PR would be:

* Keep the pie chart for consistency with the other cards for now
* In the legend, show both the actual values (GiB, TiB) and percentage in brackets, e.g. "15.0 TiB used (15 %)", "85 TiB avail. (85%)". If this makes the legend items too big, display the actual values instead the percentage.
* Display the overall capacity in the legend, below the "used" and "avail" points., e.g. "Total: 100 TiB". If that's not easily doable, we could add it to the card title instead (even though that somehow breaks consistency with the other cards). As a last resort, we could show the total amount as part of the tooltips when hovering.
* When hovering over the chart with the mouse, update the tooltip to use the same order than the chart legend, e.g. "15.0 TiB used (15 %)" (instead of the current format "Used (nn %): xx GiB"). If the total amount should also be displayed here (see previous point), the tooltip should read "xx GiB of yy TiB used (nn %)"

Does that make sense?

tspmelo commented:
Best solution I found is to show the total in the bottom of the charts:
![localhost_4200_](https://user-images.githubusercontent.com/399326/57089669-7ad35100-6cf4-11e9-804b-5bbe935550ef.png)

Should  I also add the raw value to the labels of the other charts or leave it only on the `Raw Usage`?

julienlim commented:
@tspmelo @LenzGr 

Seems like we all want to show total.  To be more consistent with the other "totals" shown in the other parts of the dashboard (see screenshot below), can we instead show it as:

33 GiB total    <---- instead of Total: 33 GiB
3 ops total

<img width="1017" alt="Screen Shot 2019-05-02 at 4 07 48 PM" src="https://user-images.githubusercontent.com/5589919/57103674-79098d80-6cf4-11e9-8bf2-95c71bdc3405.png">

Thoughts?

LenzGr commented:
> @tspmelo @LenzGr
> 
> Seems like we all want to show total, and I think underneath the pie chart seems fine. To be more consistent with the other "totals" shown in the other parts of the dashboard (see screenshot below), can we instead show it as, e.g.:
> 
> 33 GiB total <---- instead of Total: 33 GiB
> 3 ops total <---- instead of Total: 3 ops
> 107 pgs total <---- instead of Total: 107 pgs
> 
> Thoughts?

Absolutely agree, thanks for pointing this out. Consistency FTW!

tspmelo commented:
> However, I'm not quite sure about "total ops", which I assume would accumulate both reads and writes? Is there any value in this metric or isn't it actually the same value we already display in the "Client IOPS" card?

Same applies to total pgs, it is showing the same in the `PGs per OSD` card

tspmelo commented:
Remove the total from 2 of the charts, since that value is already displayed in another card.
![localhost_4200_ (2)](https://user-images.githubusercontent.com/399326/57302821-8f846000-70cb-11e9-92d2-fb7ba6183df7.png)

alfonsomthd commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/dashboard/health/health.component.ts:
```diff
@@ -79,24 +86,29 @@ export class HealthComponent implements OnInit, OnDestroy {
       100 * (data.df.stats.total_used_raw_bytes / data.df.stats.total_bytes)
     );
 
+    chart.dataset[0].label = 'asd';
```
What does 'asd' means?

tspmelo replied:
It was just some code I added to tests the labels and forgot to remove.
It has been removed, thanks for spotting.

tspmelo commented:
jenkins retest this please

epuertat commented:
Sorry to intervene in the last minute (been on PTO throughout most of this PR lifespan). What was the agreement on the pie chart vs. donut chart? Didn't know the reason, but I get the impression that UI trend is to move from pie to donuts, so just googled "pie chart vs. donut chart" and found [[1]](http://hypsypops.com/pie-chart-donut-chart/) [[2]](https://charting-ahead.corsairs.network/doughnuts-vs-pie-a-visual-primer-629b846aac67), which summarized:
- **Understanding**: Donut charts are basically folded bar charts (lengths), while pie charts rely on area-based representations. It seems that we humans are far better at understanding lengths than areas.
- **Redundancy**: Donut charts require less visual space to represent the same information as pie charts, hence they're less visually verbose/stressing.
- **Extra info**: The white space inside donut charts, which is an eye-catcher, allows for showing additional info (totals, percentages, etc).

Don't mean to change anything, just to make sure we've evaluated both options.

tspmelo commented:
Since we had more pie charts and the main reason for this PR was to have them all the same, I went with pie chart.

I tried to use donut charts with info in the center, but the chart plugin we use don't have native support for that and when I tried to add it via an extension I was not able to align the text in the center of the donut.

This PR was initially create to fix http://tracker.ceph.com/issues/39384 and only after that I added some of the changes mentioned in http://tracker.ceph.com/issues/38697.

I don't think I implemented everything 38697 suggests, so I didn't reference it in the PR. 
Maybe we can leave it open and do additional changes when someone works on them.

epuertat commented:
> I tried to use donut charts with info in the center, but the chart plugin we use don't have native support for that and when I tried to add it via an extension I was not able to align the text in the center of the donut.

Thanks! Didn't know that. I was wondering how it was done in dashboard_v1 :thinking: 

![image](https://user-images.githubusercontent.com/263427/57065693-86aa1d80-6cca-11e9-93d7-a11c2629f346.png)

tspmelo commented:
> Thanks! Didn't know that. I was wondering how it was done in dashboard_v1

It works OK when you don't have labels and only 1 small line of text. 
If you start adding more stuff is when it gets more complicated :)
